### ISSUE TITLE
Instantiating a Type[C] can typecheck but fail for abstract subtypes

### ISSUE THREAD

**Issue by bensimner** [python/mypy#9505]
Consider

```
from typing import Type
from abc import ABC, abstractmethod

class A(ABC):
    @abstractmethod
    def f(self) -> None:
        pass

class C:
    pass

class D(A, C):
    pass

def f() -> None:
    x: Type[A] = D  # Bad!  error: Can only assign concrete classes to a variable of type "Type[A]"
    y: Type[C] = D  # Ok !

OUTPUT = """\
$ python3 -m mypy --version
mypy 0.782
$ python3 -m mypy --strict test.py
test.py:16: error: Can only assign concrete classes to a variable of type "Type[A]"
Found 1 error in 1 file (checked 1 source file)
"""
```
( see https://mypy-play.net/?mypy=latest&python=3.8&gist=050667433fa08bc65f0514cdf74b95c8&flags=verbose%2Cshow-column-numbers%2Cstrict )

This at first seems to make sense, `D` is a concrete implementation of `C` but not of `A`.
But if so, then you can pass a `D` to a function `Type[C] -> C` like so:

```
from typing import Type
from abc import ABC, abstractmethod

class A(ABC):
    @abstractmethod
    def f(self) -> None:
        pass

class C:
    pass

class D(A, C):
    pass

def f(cls: Type[C]) -> C:
    return cls()

f(D)

OUTPUT = """\
$ python3 -m mypy --version
mypy 0.782
$ python3 -m mypy --strict test.py
Success: no issues found in 1 source file
$ python3 test.py
Traceback (most recent call last):
  File "test.py", line 18, in <module>
    f(D)
  File "test.py", line 16, in f
    return cls()
TypeError: Can't instantiate abstract class D with abstract methods f
"""
```

(see https://mypy-play.net/?mypy=latest&python=3.8&gist=24beec8c1b0d0d916628af2b32dd4052&flags=verbose%2Cshow-column-numbers%2Cstrict )

but doing that is a runtime error,  although mypy happily accepts it even with --strict.

I'm not sure what the "expected" outcome here is really, the first file seems reasonable and the second is a logical consequence of it.

**Comment by gvanrossum** [comment]
Illogical consequences are never fun. ;-)

Note that you don't even need multiple inheritance, you could write
```
class C(ABC): ...
class D(C):
    @abstractmethod
    def f(self): ...
def f(cls: Type[C]): cls()
f(D)
```

Mypy makes a bunch of assumptions here, and the weakest one is perhaps that the signature of all subclasses is the same as that of the superclass.

But what are we to do? Perhaps `Type[C]` should mean "you can only pass `C`, not any subclass of it", but that would not be very useful -- not useful enough anyway to warrant implementing the construct `Type[]`. :-)

Or perhaps we should require subclass constructors to be covariant with respect to the superclass constructor if the latter is used anywhere in a `Type[]` form. But that kind of global analysis is impractical. Instead, we could invent a class decorator that has two effects: (a) it requires subclass constructors to be covariant, and (b) it allows the class to be used in `Type[]`. But that seems a burden for users.

Note that it's also easy to come up with scenarios where it would be handy to accept an abstract subclass (e.g. if the class is never instantiated, but used e.g. in `isinstance()`).

So basically `Type[]` is always a bit of a compromise.

I assume there are other issues open about this, so I'm just going to close this if you don't mind. (If you insist I will reopen it, but I probably won't have time to do anything about it.)
